[Refactor] Encrypted Logging Library

ParaskP7 opened the pull request:
Closes: [AINFRA-748](https://linear.app/a8c/issue/AINFRA-748)

---

### Description

This PR is a prerequisite for [AINFRA-749](https://linear.app/a8c/issue/AINFRA-748) (`EventBus` -> `Coroutines`) and [AINFRA-748](https://linear.app/a8c/issue/AINFRA-750) (`WellSQL` -> `Room`), which prepares the ground for both those migrations. Main changes include but not limited to:

- Prettifying `libs.versions.toml` by renaming/reordering versions/libraries/plugin and extracting new versions. This includes updating the `jna` dependency, which actually solves the `is not 16 KB aligned` problem on it, leaving only `terl-lazysodium-android` with this kind of problem.
- Upgrading all gradle/plugins/libraries to their latest stable versions. This includes updating `compileSdk` (to `35`) and compile source/target compatibility, along with kotlin target to `Java 11`.
- Prettifying `build.gradle.kts` files. This includes adding/removing (unused) source sets.
- Removing the `wordpress-utils` dependency altogether. This includes copying the `DateTimeUtils` over and using Android's `Log` utility over the library's `AppLog`.

---

### Testing Steps

Using the [DOAndroid](https://github.com/bloom/DayOne-Android) or/and [PCAndroid](https://github.com/Automattic/pocket-casts-android) project, update `automattic-encryptedlogging` to `18-a7448ac51d5cf72d5974915c6cfebbd1c82c863d` (version which is based on this PR's [latest](https://github.com/Automattic/EncryptedLogging/pull/18/commits/a7448ac51d5cf72d5974915c6cfebbd1c82c863d) commit) and make sure that encrypted logging continues to work as expected:

1. Using `App Inspection` from AS find the `encrypted-logging.db` and its sole `EncryptedLogModel` table. Notice that the table is empty, there are no log to be sent.
2. Make the app crash. Using [DOAndroid](https://github.com/bloom/DayOne-Android) for example, you could navigate to `Settings` -> `Developer` and then click `Crash The App`.
3. You might now notice a new db entry within `App Inspection` from AS on that `DETACHED` app (or not, it depends). Now, switch to the `ATTACHED` app and notice the db empty. This is because this db entry/log has been already sent to Sentry already, on app start and after the app crashed. To really notice the entry, trying crashing the app again, this time put the device on offline mode first. Doing that you'll be able to see the entry on the db, even when switching to the newly created `ATTACHED` app.
4. Open [Sentry](https://a8c.sentry.io) and navigate to `Issues`. Find the project you're testing for (for example `day-one-android` or/and `pocket-casts-android`) and check the latest `debug` exceptions. Using [DOAndroid](https://github.com/bloom/DayOne-Android) for example, you could search for `Developer caused a crash!` and check the latest crash. Now navigate at the bottom and check the `uuid`. Verify it is the same `UUID` that you saw on that db entry before.
5. Open [Read Encrypted Logs](https://mc.a8c.com/encrypted-logs.php), copy-paste that `UUID` into `Log UUID`, click `View` and make sure you can read the newly uploaded and now decrypted log.

ParaskP7 committed:
Deps: Rename coreKtx version to androidx-core

FYI: This is done for consistency purposes to the rest of the projects.

ParaskP7 committed:
Deps: Rename androidx-test versions/libraries

FYI: This is done for consistency purposes to the rest of the projects.

ParaskP7 committed:
Deps: Rename appcompat version to androidx-appcompat

FYI: This is done for consistency purposes to the rest of the projects.

ParaskP7 committed:
Deps: Rename wordpress-fluxc versions/libraries

FYI: This is done for consistency purposes to the rest of the projects.

ParaskP7 committed:
Deps: Extract assertj version

FYI: This is done for consistency purposes to the rest of the projects.

ParaskP7 committed:
Deps: Extract eventbus version

FYI: This is done for consistency purposes to the rest of the projects.

ParaskP7 committed:
Deps: Extract kotlinx-coroutines version

FYI: This is done for consistency purposes to the rest of the projects.

ParaskP7 committed:
Deps: Extract robolectric version

FYI: This is done for consistency purposes to the rest of the projects.

ParaskP7 committed:
Deps: Extract volley version

FYI: This is done for consistency purposes to the rest of the projects.

ParaskP7 committed:
Deps: Extract wordpress-utils version

FYI: This is done for consistency purposes to the rest of the projects.

ParaskP7 committed:
Deps: Extract wordpress-wellsql version

FYI: This is done for consistency purposes to the rest of the projects.

ParaskP7 committed:
Deps: Rename androidLibrary plugin to android-library

FYI: This is done for consistency purposes to the rest of the projects.

ParaskP7 committed:
Deps: Rename kotlin plugins

FYI: This is done for consistency purposes to the rest of the projects.

ParaskP7 committed:
Deps: Extract automattic-publish version

FYI: This is done for consistency purposes to the rest of the projects.

ParaskP7 committed:
Deps: Sort lines alphabetically in ascending order

FYI: This is done for consistency purposes to the rest of the projects.

ParaskP7 committed:
Build: Upgrade gradle to 8.14.2

Release Notes: https://docs.gradle.org/8.14.2/release-notes.html

Command: ./gradlew wrapper --gradle-version=8.14.2
--distribution-type=all --gradle-distribution-sha256-sum=
443c9c8ee2ac1ee0e11881a40f2376d79c66386264a44b24a9f8ca67e633375f
&& ./gradlew wrapper --distribution-type=all

ParaskP7 committed:
Build: Upgrade agp to 8.10.1

Release Notes: https://developer.android.com/build/releases/

ParaskP7 committed:
Build: Upgrade androidx-appcompat to 1.7.1

Release Notes: https://developer.android.com/jetpack/androidx/releases/
appcompat#1.7.1

ParaskP7 committed:
Build: Upgrade androidx-core to 1.16.0 (compileSdk = 35)

Release Notes: https://developer.android.com/jetpack/androidx/releases/
core#1.16.0

------------------------------------------------------------------------

FYI: Also, due to the fact that 'androidx.core:core:1.16.0' and
'androidx.core:core-ktx:1.16.0' requires libraries and applications that
depend on it to compile against version 35 or later of the Android APIs,
as part of this change 'compileSdk' got update to '35' as well.

ParaskP7 committed:
Build: Upgrade androidx-test-espresso to 3.6.1

Release Notes: https://developer.android.com/jetpack/androidx/releases/
test#espresso-3.6.1

ParaskP7 committed:
Build: Upgrade androidx-test-ext to 1.2.1

Release Notes: https://developer.android.com/jetpack/androidx/releases/
test#ext.junit-1.2.1

ParaskP7 committed:
Build: Upgrade androidx-test-main to 1.6.2

Release Notes: https://developer.android.com/jetpack/androidx/releases/
test#runner-1.6.2

ParaskP7 committed:
Build: Upgrade assertj to 3.27.3

Release Notes: https://github.com/assertj/assertj/releases/
tag/assertj-build-3.27.3

ParaskP7 committed:
Build: Upgrade kotlin to 2.1.21

Release Notes: https://github.com/JetBrains/kotlin/releases/
tag/v2.1.21

ParaskP7 committed:
Build: Upgrade kotlinx-coroutines to 1.10.2

Release Notes: https://github.com/Kotlin/kotlinx.coroutines/releases/
tag/1.10.2

ParaskP7 committed:
Build: Upgrade robolectric to 4.14.1

Release Notes: https://github.com/robolectric/robolectric/releases/
tag/robolectric-4.14.1

ParaskP7 committed:
Build: Upgrade wordpress-utils to 3.15.0

Release Notes: https://github.com/wordpress-mobile/
WordPress-Utils-Android/releases/tag/3.15.0

ParaskP7 committed:
Build: Suppress deps resolution mgmt repositories on unstable api usage

ParaskP7 committed:
Build: Replace hard code 1.8 kotlin jvm target with java version 1.8

ParaskP7 committed:
Build: Use java 11 for compile source/target compatibility kotlin target

FYI: Java 17 could be used instead but most projects are still using 11
(expect for PCAndroid).

ParaskP7 committed:
Build: Add test and android test kotlin related source set config

FYI: Actually, I am not even sure that these 'sourceSets' configurations
is needed. The project seems to compile fine even without them.

PS: Also, as part of this change the unnecessary 'ExampleUnitTest.kt'
unit test and its associated unnecessary 'java' source set.

ParaskP7 committed:
Deps: Move jna and lazysodium dependencies to version catalogs

FYI: A TODO pointing to https://github.com/gradle/gradle/issues/21267
was added so that those 'aar' deps are checked again after this issue
on the Gradle side gets resolved.

PS: Also, it can be noticed that the IDE now complains on those two
dependencies nad the fact that they are not 16KB aligned:
- The native library x86_64/libjnidispatch.so
(from net.java.dev.jna:jna:5.13.0) is not 16 KB aligned
- The native library arm64-v8a/libsodium.so
(from com.goterl:lazysodium-android:5.1.0) is not 16 KB aligned

ParaskP7 committed:
Deps: Update jna to 5.17.0@aar

Release Notes: https://github.com/java-native-access/jna/blob/master/
CHANGES.md#release-5170

FYI: This update also solves the 'is not 16 KB aligned' warning for this
specific 'jna' native library, leaving only 'terl-lazysodium-android'
with this kind of a problem for this project.

ParaskP7 committed:
Refactor: Copy date time utils from wordpress util library

ParaskP7 committed:
Refactor: Remove unused code from date time utils

ParaskP7 committed:
Rename .java to .kt

ParaskP7 committed:
Kotlin: Convert date time utils to kotlin

ParaskP7 committed:
Refactor: Convert date time utils public class to internal object

ParaskP7 committed:
Refactor: Update date utc from iso 8601 and adjust usage

ParaskP7 committed:
Refactor: Update iso 8601 utc from date and adjust usage

ParaskP7 committed:
Refactor: Replace wordpress utils app log with plain android log

ParaskP7 committed:
Deps: Remove unused wordpress-utils dependency

FYI: Now that all the 'wordpress-utils' related imports are gone, this
dependency can be safely removed.

wzieba commented:
> Open [Read Encrypted Logs](https://mc.a8c.com/encrypted-logs.php), copy-paste that UUID into Log UUID, click View and make sure you can read the newly uploaded and now decrypted log.

Or use [EncryptedLoggingIntelliJ](https://github.com/Automattic/EncryptedLoggingIntelliJ) 😄

wzieba commented on encryptedlogging/build.gradle.kts:
```diff
@@ -30,13 +30,15 @@ android {
-        jvmTarget = "1.8"
+        jvmTarget = JavaVersion.VERSION_11.toString()
     }
     sourceSets["main"].java.srcDirs("src/main/kotlin")
+    sourceSets["test"].java.srcDirs("src/test/kotlin")
+    sourceSets["androidTest"].java.srcDirs("src/androidTest/kotlin")
```
```suggestion
```

I doesn't seem to be needed to add anything to `java` sourcesets. What was the reason? Compilation completed without these lines.

ParaskP7 replied:
Exactly @wzieba , that's what I added to my [commit message](https://github.com/Automattic/EncryptedLogging/pull/18/commits/886ca5c28f65cc3596e1160db9e78f0c6993ca5e) too:

> FYI: Actually, I am not even sure that these `SourceSets` configurations is needed. The project seems to compile fine even without them.

FYI: Just because you added that first `main` source set as part of [this commit](https://github.com/Automattic/EncryptedLogging/commit/dc81549b4427b8905d16425e2fb212b70ec0a4c2) I though I better keep that login, but also add the missing `test` and `androidTest` source sets too... 🤷

wzieba replied:
So if it's not useful anymore, can we remove it?

ParaskP7 replied:
Yeap, done: 0b52688e6ff155e30555d246fb04b7d51e402ef1

wzieba commented on encryptedlogging/src/main/kotlin/com/automattic/encryptedlogging/utils/DateTimeUtils.kt:
```diff
@@ -0,0 +1,45 @@
+import java.util.Locale
+import java.util.TimeZone
+
+internal object DateTimeUtils {
```
Nice 👍 I hope that in next PRs we can remove it whatsoever, or, if we decide to keep date for some reason, use kotlinx date? These manual manipulations like `iso8601date.replace("Z", "+0000").replace("+00:00", "+0000")` are strange and unnecessary complex IMO.

ParaskP7 replied:
Yea, thanks @wzieba, I did reply to you [here](https://github.com/Automattic/EncryptedLogging/pull/17#discussion_r2122996213), but this discussion didn't progress further. Maybe we can continue the discussion on that here, but yea, if we decide to keep date I agree that we better use [kotlinx.datetime](https://github.com/Kotlin/kotlinx-datetime) instead, good call! 👍

wzieba replied:
I see, I didn't understand that there's anything more to say there. If `DATE_CREATED` is only used for sorting, we can sort by synthethic primary key or use timestamp, which is easier to manage than a `Date` of any format.

ParaskP7 replied:
> ...we can sort by synthethic primary key...

Oh, this is an interesting idea, hmmm... 🤔

FYI: I'll keep this comment, and other such "refactor" comments as we progress, then apply them right after I am done with [AINFRA-749](https://linear.app/a8c/issue/AINFRA-749) and [AINFRA-750](https://linear.app/a8c/issue/AINFRA-750). I am already planning for that last and 4th PR, which will deal with all the left-overs from the previous 3.

wzieba replied:
Gotcha, thanks!

wzieba reviewed (approved):
Cool! I left a minor suggestion and a drive-by comment 👍

ParaskP7 committed:
Build: Remove unnecessary kotlin source sets configuration

PR Discussion: https://github.com/Automattic/EncryptedLogging/pull/18
#discussion_r2149437349
